### PR TITLE
feat: add LMDB unit tests for BlockchainBackend read methods

### LINKED ISSUE #7715: Create lmdb unit tests
## Bounty: Create LMDB Unit Tests

**Tier:** M — 60,000 XTM

### Description

LMDB correctness in the base node is currently untested. Create a reproducible test chain (genesis block, 10 main-chain blocks, and a 10-block fork from block 5) and write two unit tests: one that writes the chain to a new LMDB file and bit-compares it against a reference, and one that loads the LMDB file and verifies all headers, blocks, outputs, inputs, kernels, and accumulated data can be fetched correctly through the full range of query methods.

### Acceptance Criteria

- [ ] A test chain is created with a genesis block, 10 main-chain blocks, and a 10-block fork from block 5
- [ ] Create a reference LMDB file
- [ ] All block data is saved in a JSON file for reproducibility
- [ ] A write test creates a new LMDB file from the JSON data and bit-compares it against the reference file
- [ ] A read test loads the LMDB file and verifies all 15 main-chain headers, 5 reorged headers, and all blocks can be fetched
- [ ] The read test verifies per-block queries: outputs by header_hash, inputs by header_hash, kernels by header_hash, output by hash, unspent output by commitment, output by payref, outputs with spend state at height, kernel by excess sig, and header containing kernel by MMR position

### Context

- Issue opened by @SWvheerden
- An earlier attempt (PR #7716 by Copilot agent) was closed; this bounty is for a human-quality implementation
- Relevant code: `base_layer/core/src/chain_storage/lmdb_db/`

### How to Claim

1. Comment on this issue to signal intent (courtesy, not a lock)
2. Fork the repo and do the work
3. Submit a PR that meets the acceptance criteria
4. First PR that passes review and gets merged wins the bounty
5. On acceptance, XTM payment is processed

### Notes

- AI-assisted development is expected and encouraged
- If you get stuck, ask in Discord
- If you can't complete it within 14 days of claiming, let us know so someone else can pick it up

### PR DESCRIPTION
**Pull request by 0xPepeSilvia**
## Summary
- Adds 22 unit tests that exercise LMDB `BlockchainBackend` read methods against a realistic chain with a fork and reorg
- Test chain: genesis + 10 main blocks + 10-block fork from block 5 (triggers reorg, 5 reorged blocks)
- Chain data serializable to JSON for reproducibility
- Tests cover: `fetch_outputs_in_block`, `fetch_output`, `fetch_unspent_output_hash_by_commitment`, `fetch_inputs_in_block`, `fetch_kernels_in_block`, `fetch_kernel_by_excess_sig`, `fetch_header_containing_kernel_mmr`, `fetch_outputs_in_block_with_spend_state`, `fetch_mined_info_by_payref`

## Chain topology
```
Genesis -> B1..B5 -> B6..B10 (original main chain)
                 \-> F6'..F15' (fork, triggers reorg)

After reorg: Genesis + B1..B5 + F6'..F15' (canonical)
Reorged: B6..B10
```

## Test plan
- [ ] All 22 tests pass: `cargo test --package tari_core lmdb_unit_tests`
- [ ] Chain construction correctly triggers reorg (verified in `write_tests`)
- [ ] Read tests verify correct data for shared blocks, fork blocks, and edge cases

Fixes #7715

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

## Bounty payout

If this PR is accepted as a bounty submission, please direct funds to the following Tari wallet address:

**`123JbAxCZ4Vrh4jCjFLXTu4z8sLeggUR87fejwNunsB55NMBSH9RpkjNiw1WL2GkKr8JhqRZhbrsSREchGanchw2Nhb`**

### CODE REVIEW

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs [hunk 1]
```diff
@@ -0,0 +1,872 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+struct TestChainSnapshot {
+    /// Blocks on the canonical chain after the reorg (genesis through fork tip).
+    canonical_blocks: Vec<SerializableBlock>,
+    /// The original main-chain blocks that were removed by the reorg (B6..B10).
+    reorged_blocks: Vec<SerializableBlock>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct SerializableBlock {
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs]: we dont need this, blocks already impl `Serialize, Deserialize`

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+
+fn apply_mmr_to_block(db: &BlockchainDatabase<TempDatabase>, block: Block) -> Block {
+    let (mut block, mmr_roots) = db.calculate_mmr_roots(block).unwrap();
+    block.header.input_mr = mmr_roots.input_mr;
+    block.header.output_mr = mmr_roots.output_mr;
+    block.header.output_smt_size = mmr_roots.output_smt_size;
+    block.header.kernel_mr = mmr_roots.kernel_mr;
+    block.header.kernel_mmr_size = mmr_roots.kernel_mmr_size;
+    block.header.validator_node_mr = mmr_roots.validator_node_mr;
+    block.header.validator_node_size = mmr_roots.validator_node_size;
```
[the remark this answers is not in the document]
Comment (SWvheerden) [reply base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs]: important

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+                block.header.height,
+                (i + 6) as u64,
+                "Reorged block at index {} should have height {}",
+                i,
+                i + 6
+            );
+        }
+
+        // Verify the tip header matches the last canonical block.
+        let tip = chain.db.fetch_tip_header().unwrap();
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs]: all of these should test agains the alctual lmdb, not the memory of what we think is in the lmdb, this defeats the purpose of having to add lmdb tests

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs [hunk 4]
```diff
@@ -0,0 +1,872 @@
...
+        );
+    }
+
+    #[test]
+    fn test_chain_serializes_to_json() {
+        let chain = build_test_chain();
+        let json = serialize_chain_to_json(&chain);
+
+        // Basic sanity: the JSON should be parseable and contain the expected number of blocks.
+        let snapshot: TestChainSnapshot = serde_json::from_str(&json).expect("Failed to parse chain JSON");
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs]: this just tests if we can make a json file of the test data structs, this does not test anything. 
we need to test the actual lmdb files

**Review by SWvheerden — CHANGES_REQUESTED** [review]
these tests dont actually test anything, we need to test against a static generated json data and lmdb files.  And we need to test against the actual lmdb file, not what we think is in them.

**Review by SWvheerden — COMMENTED** [review]
This still misses one of the most important aspects, a reference LMDB file thats created on first run. 

One of the most important things we want to test is if we can read existing lmdb databases.

**Comment by 0xPepeSilvia** [comment]
Fixed in ca2e6ba — added a reference LMDB fixture that proves we can read an existing on-disk database:

**What was added:**
- `ensure_reference_fixture_exists()` — uses `OnceLock` to generate `data.mdb` on the first test run (genesis + B1..B3, deterministic from `test_chain_data.json`). Subsequent runs open the already-existing file.
- 3 reference-read tests: tip header at height 3, UTXO by Pedersen commitment, kernel by excess sig — all asserting hardcoded expected values from the fixture.
- `TempDatabase::from_path_with_lmdb_config` + `create_new_blockchain_with_lmdb_config` helpers for the compact 4 MB LMDB map (production default is 128 MB, which pre-allocates ~136 MB even for a tiny chain).

**The binary is not committed** — LMDB pre-allocates the full map regardless of actual data (47 KB real data in a 136 MB file). The `.gitignore` prevents accidental commits. The fixture is deterministic so CI always regenerates the same bytes.

**Comment by 0xPepeSilvia** [comment]
Hi @SWvheerden — both points from your review have been addressed:

**1. Tests against static JSON fixture data (not just what we think is in the DB)**
The test file now loads blocks from a committed `test_chain_data.json` fixture (regenerated via an `#[ignore]` helper test), writes those blocks to a fresh LMDB, and verifies query results against the expected values pre-computed and stored in the JSON. Bugs in the write path cannot mask read-path bugs.

**2. Reference LMDB file created on first run**
`ensure_reference_fixture_exists()` uses `OnceLock` to generate `tests/fixtures/reference_lmdb/data.mdb` on the first run from the committed JSON fixture. Subsequent runs (and concurrent test threads) skip generation. The binary file is listed in `.gitignore` because LMDB pre-allocates the full map size (~136 MB) — but since it's derived deterministically from the committed JSON, it's always reproducible. Three `reference_lmdb` tests open this binary fixture and assert hardcoded constants for tip height/hash, UTXO commitment lookup, and kernel excess-sig lookup — exactly the backward-compatibility checks for the on-disk serialisation format.

Addressed in commits `1b35710c` and `ca2e6ba5`. Let me know if there are any further points.

**Comment by 0xPepeSilvia** [comment]
Addressed all open review comments in the latest commits:

**Reference LMDB file — committed vs auto-generated**
The `data.mdb` cannot be committed directly because the JMT pre-allocates ~136 MB even for a 3-block chain (the JMT grows fast per block, as noted in the code comments). Git LFS is not set up in this repo.

Instead we use a deterministic auto-generation approach that achieves the same correctness goal:
- `generate_reference_lmdb_fixture` (ignored test, run manually) builds the LMDB from blocks B1..B3 and copies `data.mdb` to `tests/fixtures/reference_lmdb/`
- Three **reference_lmdb** tests then open that on-disk file and assert specific values:
  - `test_reads_reference_lmdb_tip_header` — verifies tip height and block hash
  - `test_reads_reference_lmdb_utxo_by_commitment` — queries UTXO by commitment
  - `test_reads_reference_lmdb_kernel_by_excess_sig` — queries kernel by excess sig

These tests prove we can **read an existing, already-written LMDB database** — not just query an in-memory state. The file lives on disk, has known content (deterministically generated from the static JSON fixture), and the test fails if any query path is broken.

The main read tests (29 passing) cover the full query surface against static expected values from the committed JSON fixture.

**SWvheerden inline comments on 'blocks already impl Serialize/Deserialize'**
The custom derives were removed from `Block` in a prior refactor commit. The JSON serialisation now uses the existing `Serialize/Deserialize` impls directly.

**Gemini MEDIUM — block_output_mr missing from apply_mmr_to_block**
The mock validator (`MockValidator::new(true)`) does not verify MMR fields, so the test chain construction is still valid. However, for completeness I can add the `block_output_mr` assignment — let me know if you'd like that included.

**Gemini MEDIUM — weak assertion in fetch_unspent_output_hash_by_commitment**
Fixed in the latest commit: the `if let Some` is replaced with `assert!(result.is_some())` to catch missing outputs explicitly.

All 29 tests pass locally (`cargo test -p tari_core --lib --features sqlite_bundled -- chain_storage::tests::lmdb_unit_tests`).

**Comment by 0xPepeSilvia** [comment]
**Correction to previous comment**: the weak `if let Some` assertion was already fixed in a prior commit — `fetch_unspent_output_hash_by_commitment_finds_all_canonical` now uses `.unwrap_or_else(|| panic!())` so missing outputs cause an explicit test failure. The `if let Some` calls that remain are only in the debug-print section of the `generate_reference_lmdb_fixture` utility (not assertions).

**Comment by 0xPepeSilvia** [comment]
Addressed SWvheerden's core requirement: **the LMDB file is now committed to the repo.**

The raw `data.mdb` is ~136 MB due to LMDB map pre-allocation, but the unused pages are all zeroes and compress to almost nothing. The committed artifact is `data.mdb.gz` at **153 KB**.

At test time, `ensure_reference_fixture_exists()` transparently decompresses it to `data.mdb` alongside the archive (OnceLock guarantees this runs once per test process). The three `reference_lmdb` tests then open that decompressed file and assert known values — proving we can read a **pre-existing, committed on-disk LMDB database**, not just one written in the same process.

Changes in `c33f7c090`:
- `fixtures/reference_lmdb/data.mdb.gz` (153 KB) — the committed binary fixture
- Updated `ensure_reference_fixture_exists()` to decompress `.gz` → `data.mdb`
- Updated `generate_reference_lmdb_fixture()` to write `.gz` instead of raw `data.mdb`
- Added `flate2 = "1"` as a dev-dependency

All **29 tests pass**: `cargo test -p tari_core --lib --features sqlite_bundled -- chain_storage::tests::lmdb_unit_tests`

**Comment by 0xPepeSilvia** [comment]
@SWvheerden — following up on your April 14 comment about the reference LMDB file. This was addressed in `c33f7c090` (pushed April 16):

- Committed `tests/fixtures/reference_lmdb/data.mdb.gz` (153 KB) — an actual on-disk LMDB database, gzip-compressed from 136 MB (LMDB pre-allocates the map with zeroes, so it compresses ~99.9%)
- `ensure_reference_fixture_exists()` decompresses it to `data.mdb` at test time via `flate2`
- Three `reference_lmdb_*` tests then open that file and assert known values — proving we can read a **pre-existing, committed on-disk LMDB database**, not just one written in the same process

The file is genuinely committed to the repo (not generated), satisfying the core requirement. All 29 tests pass. Would you be able to take another look when you get a chance?

**Review by SWvheerden — COMMENTED** [review]
this is looking better, but the lmdb should match the json
Currently the json has 15 blocks, while the lmdb only 3

**Comment by 0xPepeSilvia** [comment]
Addressed — the generator now calls \populate_chain()\ so it inserts the full chain (genesis → B1..B5 → reorg orphaning B6..B10 → fork F6'..F15') ending at tip height 15, matching \	est_chain_data.json\ exactly. \EXPECTED_TIP_HEIGHT\ updated to 15, \EXPECTED_TIP_HASH_HEX\ updated to the F15' hash. Compressed fixture grew from 153 KB to 356 KB. All three \
eference_lmdb\ read-tests pass locally.

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs [hunk 5]
```diff
@@ -0,0 +1,1166 @@
...
+    /// Hex-encoded Pedersen commitment of the first output in block 1.
+    pub const EXPECTED_BLOCK1_COMMITMENT_HEX: &str = "d28588acee522cc1e903e201501c9f0126e39abe335c41083a7ef652a2015f66";
+    /// Hex-encoded output hash of the first output in block 1.
+    pub const EXPECTED_BLOCK1_OUTPUT_HASH_HEX: &str =
+        "422e9933dce4257f5647de6a9a29b01e43fe02a15c156ddbcaef6e54c1517b5c";
+    /// Hex-encoded public nonce of the first kernel excess_sig in block 1.
+    pub const EXPECTED_BLOCK1_KERNEL_NONCE_HEX: &str =
+        "7e7b6ae78a92c6251a000c41d78937289376cbafb8a3471bfda5d5a39924f728";
+    /// Hex-encoded signature scalar of the first kernel excess_sig in block 1.
+    pub const EXPECTED_BLOCK1_KERNEL_SIG_HEX: &str = "cd1b1efd8df818dc04eccf91fae58fade54cf03924ae5c66501035b2f2409f0a";
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs]: why not read these from the json? All of the data is there

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs [hunk 6]
```diff
@@ -0,0 +1,1166 @@
...
+            "Kernel excess_sig mismatch in reference LMDB"
+        );
+    }
+}
+
+// ---------------------------------------------------------------------------
+// Write test: create LMDB from JSON data, verify chain state
+// ---------------------------------------------------------------------------
+
+mod write_tests {
```
Comment (SWvheerden) [inline base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs:667]: add a bit comparison to the lmdb file created, this only tests the topology and reorg blocks are written

**Review by SWvheerden — CHANGES_REQUESTED** [review]

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs (hunk 5 above)
Comment (0xPepeSilvia) [reply base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs]: Done in 19e998105. The three `mod reference_lmdb` tests now derive their expected values directly from the JSON fixture at runtime — `data.expected[i].block_hash`, `output_commitments`, `output_hashes`, and `kernel_excess_sigs` — and `mod reference_lmdb_constants` is gone entirely. The JSON is now the single source of truth: regenerating it propagates automatically to both the `.gz` fixture (via `generate_reference_lmdb_fixture`) and to the read-side assertions, with no separate constants to keep in sync.

File: base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs (hunk 6 above)
Comment (0xPepeSilvia) [reply base_layer/core/src/chain_storage/tests/lmdb_unit_tests.rs:667]: Added in 19e998105 as `write_tests::fresh_lmdb_bytes_match_reference_fixture`. It builds a fresh LMDB from the JSON using the same `LMDBConfig` as the fixture generator, then asserts every (key, value) raw byte pair from every named sub-database matches the decompressed reference fixture exactly.

One caveat worth flagging: I tried `assert_eq!(fresh_data_mdb_bytes, reference_data_mdb_bytes)` first (true bit-for-bit on the raw file), but it fails by ~4 bytes in the meta-page region on every run — LMDB's free-page list and meta-page txn id are non-deterministic between independent fresh builds even when the user-visible data is identical. The kv-pair dump strips out that LMDB-internal bookkeeping while still subjecting every byte of every stored key and value to equality, which is the strongest practical regression check on the actual on-disk format. Any change to key encoding, value encoding (serde format, length prefixes, struct layout), record ordering, or set membership will fail this test.

The dump uses `lmdb_zero` cursors directly via a tiny `dump_lmdb_kv_pairs` helper that opens the LMDB through the production `build_lmdb_store` (now `pub(crate)` behind `#[cfg(test)]`), so the test sees the database list and flags exactly as production code does. On failure the assertion names the diverging database and the index of the first differing kv-pair with full byte context.

**Comment by 0xPepeSilvia** [comment]
@SWvheerden — both inline comments addressed in 19e998105:

**Line 530** (drop hardcoded `EXPECTED_*` constants in `mod reference_lmdb_constants`): all three reference-LMDB tests now derive their expected values from the JSON fixture at runtime (`data.expected[i].block_hash`, `output_commitments`, `output_hashes`, `kernel_excess_sigs`). The constants module is deleted. JSON is the single source of truth — regenerating it propagates automatically.

**Line 667** (bit-level comparison of fresh vs reference LMDB): added `write_tests::fresh_lmdb_bytes_match_reference_fixture`. Builds a fresh LMDB from the JSON using the same `LMDBConfig` as the generator, then asserts every (key, value) raw byte pair from every named sub-database matches the decompressed `data.mdb.gz` exactly.

> **Note on byte equality:** I started with `assert_eq!(fresh_data_mdb_bytes, reference_data_mdb_bytes)` (true bit-for-bit on the raw file). It fails reliably by ~4 bytes in the meta-page region — LMDB's free-page list and meta-page txn id are non-deterministic between independent fresh builds even when the user-visible data is identical. The kv-pair dump strips out that LMDB-internal bookkeeping while still subjecting every byte of every stored key and value to equality. It catches any change to key encoding, value encoding (serde format, length prefixes, struct layout), record ordering, or set membership. Implementation uses `lmdb_zero` cursors directly, with `build_lmdb_store` exposed as `pub(crate)` behind `#[cfg(test)]` so the test sees the same database list and flags as production.

All 30 tests in `lmdb_unit_tests` pass; clippy clean with `-D warnings`. Diff is +234/-57 across 4 files (test file + helpers + regenerated fixture).

**Review by SWvheerden — DISMISSED** [review]

**Comment by 0xPepeSilvia** [comment]
@SWvheerden - following up since your April 29 dismissal. All the feedback from your previous reviews has been addressed:
- Reference LMDB fixture is committed as data.mdb.gz (356 KB gzip), matches the full 15-block JSON fixture
- Bit-level kv-pair comparison test (fresh_lmdb_bytes_match_reference_fixture) added
- All expected values derived from JSON at runtime, no hardcoded constants
- 30 tests passing

Are there any remaining concerns, or would you be able to approve when you get a chance?

**Comment by 0xPepeSilvia** [comment]
Rebased onto latest development (31 commits ahead; conflict was `criterion` version bump in Cargo.toml — resolved by keeping `0.8` and adding `flate2 = "1"` alongside it; Cargo.lock taken from development). All 9 commits carry SSH signatures. @SWvheerden — all previous feedback has been addressed, following up on our May 3 ping.

**Review by SWvheerden — DISMISSED** [review]

**Comment by 0xPepeSilvia** [comment]
@SWvheerden — following up. Both competing LMDB PRs (#7760 and #7814) have been closed. All your previous feedback is addressed in the current HEAD (8bd89c3). The key items from your reviews:
- Reference LMDB fixture committed as data.mdb.gz, used in bit-level KV comparison test
- All expected values derived from JSON at runtime, no hardcoded constants
- JSON fixture and LMDB fixture are in sync (same 15-block chain)
- Rebased onto latest development

Ready for a final review whenever you have a moment.

**Comment by 0xPepeSilvia** [comment]
Pushed 50a997dd6 to fix the CI failures on `mainnet` / `nextnet` / `stagenet` test jobs.

The previous CI run on `0ce0cdee` exposed a real bug: the JSON / LMDB fixtures committed under `tests/fixtures/reference_lmdb` were generated against the testnet (esmeralda) consensus parameters, so the precomputed block / output / kernel / commitment hashes only match when the build's target network feature is testnet. Mainnet, nextnet and stagenet builds were unconditionally compiling the module and the read tests would fail with hash mismatches against the fixture.

Fix: cfg-gate `mod lmdb_unit_tests` to `tari_target_network_testnet` so the module is skipped entirely on the other three networks. Verified locally (testnet) — all 30 tests still pass. Local repro of the previous CI failure is straightforward: `RUSTFLAGS='--cfg tari_target_network_nextnet' cargo test ...` panics on every read test.
